model/src/id.rs

```diff
@@ -14,19 +14,20 @@ pub(crate) mod string {
     use std::{
         fmt::{Display, Formatter, Result as FmtResult},
         marker::PhantomData,
+        num::NonZeroU64,
     };
 
-    struct IdVisitor<T: From<u64>>(PhantomData<T>);
+    struct IdVisitor<T: From<NonZeroU64>>(PhantomData<T>);
 
-    impl<'de, T: From<u64>> Visitor<'de> for IdVisitor<T> {
+    impl<'de, T: From<NonZeroU64>> Visitor<'de> for IdVisitor<T> {
         type Value = T;
 
         fn expecting(&self, f: &mut Formatter<'_>) -> FmtResult {
             f.write_str("string or integer snowflake")
         }
 
         fn visit_u64<E: DeError>(self, value: u64) -> Result<Self::Value, E> {
-            Ok(T::from(value))
+            Ok(T::from(NonZeroU64::new(value).expect("not zero")))
```

Comment:
We should return a [`serde::de::Error::unexpected_value`](https://docs.rs/serde/1.0.127/serde/de/trait.Error.html#method.invalid_value) here instead of panicking. IIRC this should look something like:
```suggestion
            use serde::de::{Error as DeError, Unexpected};

            NonZeroU64::new(value).map(T::from).ok_or_else(|| {
                let unexpected = Unexpected::Unsigned(value);

                DeError::invalid_value(unexpected, &"a natural number")
            })
```
```

Reply:
The natural numbers include zero so I changed the error message to "a positive integer"

Reply:
In some definitions zero is indeed a natural number, so strictly speaking its status can't be definitively stated. Zero is neither definitively a positive nor a negative, so its status can't be definitively stated either. One could say "non-negative" but it has the same problem. Let's just go with "expected an integer >= 1"